types/rosie/index.d.ts

```diff
@@ -8,229 +8,267 @@
+         * conflict.
+         *
+         *   Factory.define('Person').sequence('id');
+         *
+         * @param {string} attr
+         * @param {Array.<string>=} | Function dependencies or builder
+         * @param {function(number): *=} builder
+         * @return {Factory}
+         */
+        sequence<K extends keyof T>(name: K, builder?: (i: number) => any): IFactory<T>;
+        sequence<K extends keyof T, D extends keyof T>(
+            name: K,
+            dependencies: D[],
+            builder: (i: number, ...args: any[]) => any,
+        ): IFactory<T>;
+
+        /**
+         * Sets a post-processor callback that will receive built objects and the
+         * options for the build just before they are returned from the #build
+         * function.
+         *
+         * @param {function(object, ?object)} callback
+         * @return {Factory}
+         */
+        after(functionArg: (obj: T, opts?: any) => void): IFactory<T>;
+
+        /**
+         * Sets the constructor for this factory to be another factory. This can be
+         * used to create more specific sub-types of factories.
+         *
+         * @param {Factory} parentFactory
+         * @return {Factory}
+         */
+        inherits(functionArg: (parentFactory: IFactory<T>) => void): IFactory<T>;
+
+        /**
+         * Builds a plain object containing values for each of the declared
+         * attributes. The result of this is the same as the result when using #build
+         * when there is no constructor registered.
+         *
+         * @param {object=} attributes
+         * @param {object=} options
+         * @return {object}
+         */
+        attributes(attributes?: { [k in keyof T]?: T[k] }, options?: any): T;
+
+        /**
+         * Generates values for all the registered options using the values given.
+         *
+         * @private
+         * @param {object} options
+         * @return {object}
+         */
+        options(options: any): any;
+
+        /**
+         * Builds objects by getting values for all attributes and optionally passing
+         * the result to a constructor function.
+         *
+         * @param {object=} attributes
+         * @param {object=} options
+         * @return {*}
+         */
+        build(attributes?: { [k in keyof T]?: T[k] }, options?: any): T;
+
+        buildList(size: number, attributes?: { [k in keyof T]?: T[k] }, options?: any): T[];
+
+        /**
+         * Extends a given factory by copying over its attributes, options,
+         * callbacks, and constructor. This can be useful when you want to make
+         * different types which all share certain attributes.
+         *
+         * @param {string|Factory} name The factory to extend.
+         * @return {Factory}
+         */
+        extend<K extends T>(name: string): IFactory<K>;
+        extend<K extends T, T>(name: IFactory<T>): IFactory<K>;
```

Comment:
It seems like the problem is that this `T` hides the one from the containing type.  I think you can just rename it, rather than adding an overload:
```ts
extend<K extends T, Z>(name: String | IFactory<Z>): IFactory<K>;
```
I have no idea what it represents, so you'll have to come up with a more sensible name. :wink:

Reply:
@amcasey Thank you for your kindly response!

It's working what you suggested, but I only have a red flag with that:

`T` is defined in the main IFactory definition (which defines the main model that rosie (a flexible factory) will instantiate).

With that in mind, if you can give an unscoped `Z`, it will be potential problem.

Thinking about this (and I have to check with the source code of the lib) the most approximate implementation would be:

```
extend<K extends Partial<T>>(name: IFactory<K> | string): IFactory<T>;
```

You want to extend from T or a superclass of T, and always return the same type. Else you might be introducing a bug.

I added a few more tests to illustrate what the lib is doing with that method.

@amcasey  I'm not finding a way to check that K is a supertype of T. The partial make it works but also introduces a bug which I added as a commented test just to bypass the ci while while we discuss this topic.

Reply:
The `T` in your original example was already unconstrained, so I was just following that.  I'm happy to help you think through what you want to express, but I'll need some more examples.

```ts
({} as IFactory<Dog>).extend({} as IFactory<Animal>) // IFactory<Dog>, IFactory<Animal>, or illegal?
({} as IFactory<Dog>).extend({} as IFactory<Dachshund>) // IFactory<Dog>, IFactory<Dachshund>, or illegal?
```

Does `extend` mutate the original instance (i.e. the receiver) or make a new one?  If it's just mutating, it seems like you might want to return `this`?